packages/svelte2tsx/src/nodes/ComponentEvents.ts

```diff
@@ -0,0 +1,82 @@
+import ts from 'typescript';
+import { EventHandler } from './event-handler';
+
+export abstract class ComponentEvents {
+    protected events = new Map<string, { type: string; doc?: string }>();
+
+    getAll(): { name: string; type?: string; doc?: string }[] {
+        const entries: { name: string; type: string; doc?: string }[] = [];
+
+        const iterableEntries = this.events.entries();
+        for (const entry of iterableEntries) {
+            entries.push({ name: entry[0], ...entry[1] });
+        }
+
+        return entries;
+    }
+
+    abstract toDefString(): string;
+}
+
+export class ComponentEventsFromInterface extends ComponentEvents {
+    constructor(node: ts.InterfaceDeclaration) {
+        super();
+        this.events = this.extractEvents(node);
+    }
+
+    toDefString() {
+        return '{} as unknown as ComponentEvents';
+    }
+
+    private extractEvents(node: ts.InterfaceDeclaration) {
+        const map = new Map<string, { type: string; doc?: string }>();
+
+        node.members.filter(ts.isPropertySignature).forEach((member) => {
+            map.set(member.name.getText(), {
+                type: member.type?.getText() || 'Event',
+                doc: this.getDoc(node, member),
+            });
+        });
+
+        return map;
```

Comment:
The name here is possibly a string literal or a computed property. i.e.
```
interface ComponentEvents {
    'kebab-case': CustomEvent<number>
    [SOME_CONSTANT]: CustomEvent<number>
}
```
`getText` in either situation would be wrong. if it's a string literal the text can be extracted from `member.name.text`.
About the computed property, from my testing, computed event handler like <Component on:[SOME_CONSTANT] /> doesn't work. Maybe we could make it works here but filter out in the completion. Or simply doesn't support it as it doesn't make much sense when you can't use it in markup.

Reply:
Good catch! We should definitely support string literals. About the constant, we could go to the root and check if it's defined there and get the value from there - so we at least support it a little. I will also add docs about this.